accepted/2020/binary-data/binary-data.md

```diff
@@ -0,0 +1,150 @@
+# Type for holding & converting binary data
+
+**PM** [Immo Landwerth](https://github.com/terrajobst) |
+[GitHub Issue](https://github.com/dotnet/runtime/issues/41686)
+
+In the BCL we have many representations for binary data such as `byte[]`,
+`Span<byte>`, and `Memory<byte>`. This isn't helped by they fact that there are
+also multiple ways how one can produce a binary payload for a `string` or
+`object`.
+
+The Azure SDK has made the experience in user studies that customers are often
+struggling to find the right method, or even any, to produce the data from the
+data they already have.
+
+## Scenarios and User Experience
+
+### Sending a binary payload
+
+Naomi is building a cloud-based solution for batch resizing of images. She uses
+Azure Storage Queues to manage requests for image resizing. In order to queue a
+message, she uses the `SendMessageAsync` API. The method takes a type called
+`BinaryData`. Curious what this is, she decides to new one up. Since there is a
+constructor that takes an object and it works for her:
+
+```C#
+Task QueueImageResize(string imageUrl)
+{
+    var queueClient = new QueueClient(userSecret.QueueConnectionString, "resize-queue");
+
+    var message = new ResizeImageMessage
+    {
+        Url = imageUrl,
+        Resolutions = new[] {"200x200", "400x400", "1200x1200" }
+    };
+
+    var messageData = new BinaryData(message);
+    await queueClient.SendMessageAsync(messageData);
+}
+```
+
+### Retrieving a binary payload
+
+Shirin works on Naomi's team but is responsible for the resizing backend. She
+wrote an Azure Function that is triggered every time an item is added to the
+queue. The template already created a handler `RunAsync` that passes the queue
+item as an instance of `BinaryData`.
+
+In order to deserialize, she simply invokes the `ToObjectFromJson` method:
+
+```C#
+Task RunAsync(BinaryData messageData)
+{
+    var message = messageData.ToObjectFromJson<ResizeImageMessage>();
+
+    foreach (var resolution in message.Resolutions)
+    {
+        var newUrl = GetUrl(resolution.Url, resolution);
+        var image = await DownloadImage(resolution.Url);
+        image.Resize(resolution);
+        await UploadImageAsync(newUrl, image);
+    }
+}
+```
+
+## Requirements
+
+### Goals
+
+* Stable API
+* Make it easy for people to construct bytes from common data types (byte
+  arrays, spans, memories, string, JSON serialized objects)
+* Make it easy for people to convert bytes to common data types of the same
+  representations
+* Provide opinionated defaults in order to reduce complexity
+    - String encoding is assumed to be UTF8
+    - Object serialization is done with `System.Text.Json`, with specific options
```

Comment:
I realize this may need a larger discussion, but have we considered porting over something like `ObjectSerializerBase` as a serialization abstraction into .NET as well? We could then have the Azure.Core [ObjectSerializer](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Serialization/ObjectSerializer.cs) inherit from it, and we would be able to include the `ObjectSerializer` APIs on `BinaryData` as we have currently on the [type ](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs#L251) that is in preview in Azure.

Reply:
Yes, we talked about this and this seems like a good idea. I just filed https://github.com/dotnet/runtime/issues/43669 to track this. We only excluded this from this feature for expediency.

Reply:
That's great news! 

One point to clarify - say we end up adding in `ObjectSerializer` (or maybe `ObjectSerializerBase` since the shipped version in Azure.Core is called `ObjectSerializer`) after shipping `BinaryData`. If we then want to expose the `ObjectSerializer` within `BinaryData` in `ToObject/FromObject` methods, would this change how we think about the `FromObjectAsJson/ToObjectAsJson` methods?  `FromObject/ToObject` would be able to be used with Json as well if you pass in an ObjectSerializer instance that uses Json, e.g. the [JsonObjectSerializer](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/Serialization/JsonObjectSerializer.cs), so it might make more sense to name all methods as FromObject/ToObject and indicate that the default behavior is Json through the parameter name as we [currently do](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs#L235). Is this something we would consider changing after the initial release of BinaryData?

Reply:
> Is [renaming ToObjectAsJson] something we would consider changing after the initial release of BinaryData?

No, once it's stable (especially as a System.* namespaced type) it's that way basically forever.

> If we then want to expose the ObjectSerializer within BinaryData in ToObject/FromObject methods, would this change how we think about the FromObjectAsJson/ToObjectAsJson methods?

I don't think so.  The ones that take the configured serializer do whatever the provided serializer instance does. The ones that don't take it provide a set of ease-of-use defaults.  E.g. if both JSON and protobuf were top-level ease-of-use constructs you can't have

```C#
public byte[] ToObject<T>(object jsonSerializable, JsonSerializerOptions options = default);
public byte[] ToObject<T>(object protobufSerializable, ProtobufSerializerOptions options = default);
```

Well, you can, but then it's a compile error for any to call `ToObject(obj)` since there are two valid targets (so if they're done across two different versions then the Protobuf method introduced a breaking change; if they're in the same version it's as if neither had defaults).  Providing the data format name in the method allows for defaulted options to be used across all formats.  (This is also true for the constructors, a protobuf-based ctor can't have defaulted options, so that format might be method-only... but since it's in a V2+ of the type there is already a body of people to spread the awareness of the factory methods).

> or maybe `ObjectSerializerBase` since the shipped version in Azure.Core is called `ObjectSerializer`

There's a rule for this :smile::
* AVOID naming base classes with a “Base” suffix if the class is intended for use in public APIs.

But the naming and scope of that type/feature is a different issue.